### PR TITLE
feat(abi): Tuples as inputs/outputs to main

### LINKED ISSUE #2891: Array of tuple not supported in the main function
### Aim

I tried to pass array of tuple to the main function

### Expected Behavior

It should have compiled successfully 

### Bug

```
Message:  not yet implemented: AbiType::from_type not yet implemented for tuple types
Location: tooling/noirc_abi/src/lib.rs:167
```

### To Reproduce

```
global DEPTH: Field = 11;
fn main(x: [(u8, u8, u8); DEPTH], y: u8) {}
```

Compile this

### Installation Method

None

### Nargo Version

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

Yes

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Resolves #2891, #1855

## Summary\*

This PR adds basic support for tuple inputs/outputs to main. 

## Documentation

- [X] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [X] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: tooling/noirc_abi/src/input_parser/json.rs [hunk 1]
```diff
@@ -104,6 +104,13 @@ impl JsonTypes {
                 JsonTypes::Table(map_with_json_types)
             }
 
+            (InputValue::Vec(vector), AbiType::Tuple { fields }) => {
+                let fields = try_vecmap(vector.iter().zip(fields.iter()), |(value, typ)| {
```
Comment (TomAFrench) [inline tooling/noirc_abi/src/input_parser/json.rs:108]:
```suggestion
                let fields = try_vecmap(vector.iter().zip(fields), |(value, typ)| {
```

File: tooling/noirc_abi/src/input_parser/json.rs [hunk 2]
```diff
@@ -169,6 +176,14 @@ impl InputValue {
                 InputValue::Struct(native_table)
             }
 
+            (JsonTypes::Array(array), AbiType::Tuple { fields }) => {
+                let tuple_fields =
+                    try_vecmap(array.into_iter().zip(fields.iter()), |(value, typ)| {
```
Comment (TomAFrench) [inline tooling/noirc_abi/src/input_parser/json.rs:181]:
```suggestion
                    try_vecmap(array.into_iter().zip(fields), |(value, typ)| {
```

File: tooling/noirc_abi/src/input_parser/mod.rs [hunk 3]
```diff
@@ -64,6 +64,17 @@ impl InputValue {
...
             }
 
+            (InputValue::Vec(vec_elements), AbiType::Tuple { fields }) => {
+                if vec_elements.len() != fields.len() {
+                    return false;
+                }
+
+                vec_elements
+                    .iter()
+                    .zip(fields.iter())
```
Comment (TomAFrench) [inline tooling/noirc_abi/src/input_parser/mod.rs:74]:
```suggestion
                    .zip(fields)
```

File: tooling/noirc_abi/src/input_parser/toml.rs [hunk 4]
```diff
@@ -102,6 +102,13 @@ impl TomlTypes {
                 TomlTypes::Table(map_with_toml_types)
             }
 
+            (InputValue::Vec(vector), AbiType::Tuple { fields }) => {
+                let fields = try_vecmap(vector.iter().zip(fields.iter()), |(value, typ)| {
```
Comment (TomAFrench) [inline tooling/noirc_abi/src/input_parser/toml.rs:106]:
```suggestion
                let fields = try_vecmap(vector.iter().zip(fields), |(value, typ)| {
```

File: tooling/noirc_abi/src/input_parser/toml.rs [hunk 5]
```diff
@@ -156,6 +163,14 @@ impl InputValue {
                 InputValue::Struct(native_table)
             }
 
+            (TomlTypes::Array(array), AbiType::Tuple { fields }) => {
+                let tuple_fields =
+                    try_vecmap(array.into_iter().zip(fields.iter()), |(value, typ)| {
```
Comment (TomAFrench) [inline tooling/noirc_abi/src/input_parser/toml.rs:168]:
```suggestion
                    try_vecmap(array.into_iter().zip(fields), |(value, typ)| {
```

File: tooling/noirc_abi/src/lib.rs [hunk 6]
```diff
@@ -370,6 +379,11 @@ impl Abi {
                     encoded_value.extend(Self::encode_value(object[field].clone(), typ)?);
                 }
             }
+            (InputValue::Vec(vec_elements), AbiType::Tuple { fields }) => {
+                for (i, typ) in fields.iter().enumerate() {
+                    encoded_value.extend(Self::encode_value(vec_elements[i].clone(), typ)?);
+                }
```
Comment (TomAFrench) [inline tooling/noirc_abi/src/lib.rs:385]:
```suggestion
                for (value, typ) in vec_elements.into_iter().zip(fields) {
                    encoded_value.extend(Self::encode_value(value, typ)?);
                }
```

**Comment by TomAFrench** [comment]
Can you reflect this in `noirc_abi_wasm`? We need to remove the duplication of this code in there but for now we need to keep these files in sync.

**Comment by vezenovm** [comment]
> Can you reflect this in `noirc_abi_wasm`? We need to remove the duplication of this code in there but for now we need to keep these files in sync.

Done

**Review by TomAFrench — APPROVED** [review]
LGTM

**Pull request merged**